docs/docs/configuration.md
# Configuration

Settings for dkron can be specified in three ways: Using a `config/dkron.json` config file, using env variables starting with `DKRON_` or using command line arguments.

## Command line options

* `-node` - Name of the node, must be unique in the cluster. By default this is the hostname of the machine.

* `-bind` - The address that dkron will bind to for communication with other dkron nodes. By default this is "0.0.0.0:8946". dkron nodes may have different ports. If a join is specified without a port, we default to locally configured port. dkron uses both TCP and UDP and use the same port for both, so if you have any firewalls be sure to allow both protocols. If this configuration value is changed and no port is specified, the default of "8946" will be used.

* `-join` - Address of another agent to join upon starting up. This can be specified multiple times to specify multiple agents to join. If Dkron is unable to join with any of the specified addresses, agent startup will fail. By default, the agent won't join any nodes when it starts up.

* `-advertise` - The advertise flag is used to change the address that we advertise to other nodes in the cluster. By default, the bind address is advertised. However, in some cases (specifically NAT traversal), there may be a routable address that cannot be bound to. This flag enables gossiping a different address to support this. If this address is not routable, the node will be in a constant flapping state, as other nodes will treat the non-routability as a failure.

* `-http-addr` - The address where the web UI will be binded. By default `:8080`

* `-backend` - Backend storage to use, etcd, consul or zk (zookeeper). The default is etcd.

* `-backend-machine` - Backend storage servers addresses to connect to. This flag can be specified multiple times. By default `127.0.0.1:2379`

* `-tag` - The tag flag is used to associate a new key/value pair with the agent. The tags are gossiped and can be used to provide additional information such as roles, ports, and configuration values to other nodes. Multiple tags can be specified per agent. There is a byte size limit for the maximum number of tags, but in practice dozens of tags may be used. Tags can be changed during a config reload.

* `-server` - If this agent is a dkron server, just need to be present. Absent by default.

* `-keyspace` - Keyspace to use for the store. Allows to run different instances using the same storage cluster. `dkron` by default.

* `-encrypt` - Key for encrypting network traffic. Must be a base64-encoded 16-byte key.

* `-mail-host` - Mail server host address to use for notifications.

* `-mail-port` - Mail server port.

* `-mail-username` - Mail server username used for authentication.

* `-mail-password` - Mail server password to use.

* `-mail-from` - From email address to use.

* `-webhook-url` - Webhook url to call for notifications.

* `-webhook-payload` - Body of the POST request to send on webhook call.

* `-webhook-header` - Headers to use when calling the webhook URL. Can be specified multiple times.

* `-log-level` - Set the log level (debug, info, warn, error, fatal, panic). Defaults to "info"

* `-rpc-port` - The port that Dkron will use to bind for the agent's RPC server, defaults to `6868`. The RPC address will be the bind address.

# Example

```json
# Dkron example configuration file
{
  "backend": "etcd",
  "backend_machine": "127.0.0.1:2379",
  "advertise_addr": "192.168.50.1",
  "server": false,
  "debug": false,
  "tags": {
    "role": "web",
    "datacenter": "east"
  },
  "keyspace": "dkron",
  "encrypt": "a-valid-key-generated-with-dkron-keygen",
  "join": [
    "10.0.0.1",
    "10.0.0.2",
    "10.0.0.3"
  ],
  "webhook_url": "https://hooks.slack.com/services/XXXXXX/XXXXXXX/XXXXXXXXXXXXXXXXXXXX",
  "webhook_payload": "payload={\"text\": \"{{.Report}}\", \"channel\": \"#foo\"}",
  "webhook_headers": "Content-Type:application/x-www-form-urlencoded",
  "mail_host": "email-smtp.eu-west-1.amazonaws.com",
  "mail_port": 25,
  "mail_username": "mailuser",
  "mail_password": "mailpassword",
  "mail_from": "cron@example.com"
}
```


CHANGELOG.md
## 0.9.4 (2017-06-07)

- Fix mistakes in API docs
- Using "jobs", "1am" or "1pm" in the name of job leads to a dashboard bug
- Fix crash on non existent plugin name
- Embed all assets in binary, removed -ui-dir config param

### Upgrade notes

This is a breaking change; `ui-dir` configuration param has been removed, all scripting using this param should be updated.

## 0.9.3 (2017-02-12)

- Fix RPC server listen address (@firstway)
- Basic implementation of the testing infrastructure using swarm
- Basic Telemetry implementation for sending metrics to statsd and datadog
- Fix crash on backend failure
- Reverse sort executions in UI (@Eyjafjallajokull)

## 0.9.2 (2016-12-28)

Features:

- Implement concurrency policy
- Improved UI: allow delete jobs from UI, highlight JSON
- Execution Processor plugins, allows flexible routing of execution results
- Template variables for customization of notification emails (@oldmantaiter)
- Go 1.7
- Test with docker-compose, this will allow to test multiple stores easily

Bug fixes:

- Fix tests randomly failing (@oldmantaiter)
- Return empty list when no jobs instead of null
- Allow POST usage on /leave method, deprecate GET

## 0.9.1 (2016-09-24)

Bug fixes:

- Fix job stats not being updated #180
- Fix zookeeper get list of executions #184
- Fix crash when deleting a job that doesn't exists #182
- Fix Travis in forks

## 0.9.0 (2016-08-24)

Features:

- Support any size jobs
- Support chained jobs
- Schedule and other job properties validation
- New site, logo and dashboard design

Bug fixes:

- Fix execution retries
- Fix executions merge by same prefix
- Fix correct HTTP status on create/update

## 0.7.3 (2016-07-12)

- One off jobs
- Added cron spec to docs
- Execution retry on failure
- Switch JSON schema spec for it's corresponding Open API spec
- Reload config
- Fix scheduling bug
- New job Status gives more info on the job execution

## 0.7.2 (2016-06-01)

- Add some helpers and bugfixes
- Add shell property to job, reintroduced the shell execution method but now it's a choice
- Add reporting node to execution reports
- Replace server tag for dkron_server and add dkron_version

### Upgrade notes

Due to the change in the internal tags `server` to `dkron_server`, you'll need to adjust job tags if they where using that tag.

## 0.7.1 (2016-05-03)

- Don't use shell call when executing commands, exploding the command line.
- Add advertise, add `advertise` option that solves joining between hosts when running docker
- Validate job size, limit to serf maximum size
- Job overwrite, now sending existing jobs doesn't overwrite non existing fields in request
- Fix for dashboard crash on non existent leader

## 0.7.0 (2016-04-10)

- Refactor leader election, the old method could lead to cases where 2 or more nodes could have the scheduler running without noticing the other master.
- Get rid of `keys`, in a serf cluster node names are unique so using it for leader keys now.
- Fix [#85](https://github.com/victorcoder/dkron/issues/85) Restart scheduler on job deletion
- Refactor logging, replace `debug` with `log-level`
- Order nodes in UI [#81](https://github.com/victorcoder/dkron/issues/81) (kudos @whizz)
- Add exposed vars to easy debugging
- Go 1.6
- Add @minutely as predefined schedule (kudos @mlafeldt)

### Upgrade from 0.6.x

To upgrade an existing installation you must first delete the pre-exiting leader key from the store. The leader key is in the form of: `[keyspace]/leader`

## 0.6.4 (2016-02-18)

- Use expvars to expose metrics
- fix https://github.com/victorcoder/dkron/issues/71
- Better example config in package and docs

## 0.6.3 (2015-12-28)

- UI: Better job view
- Logic to store only the last 100 executions

## 0.6.2 (2015-12-22)

- Fixed [#62](https://github.com/victorcoder/dkron/issues/55)

## 0.6.1 (2015-12-21)

- Fixed bugs [#55](https://github.com/victorcoder/dkron/issues/55), [#52](https://github.com/victorcoder/dkron/issues/52), etc.
- Build for linux arm

## 0.6.0 (2015-12-11)

- Some other improvements and bug fixing
- Vendoring now using Go vendor experiment + glide
- Fix: Remove executions on job delete
- Show full execution output in UI modal
- New executions results internals using RPC
- Standarized logging
- Show job tooltips with info
- Accept just "pretty" for formatting api requests
- Change how execution groups work to not use the directory concept.

## 0.5.5 (2015-11-19)

- More backend compatibility
- Accept just pretty for formatting api requests
- Show executions grouped in web UI
- Show job tooltips with all job JSON info in web UI
- Better alerts

## 0.5.4 (2015-11-17)

- Fix to web UI paths

## 0.5.3 (2015-11-16)

- Web UI works behind http proxy

## 0.5.2 (2015-11-09)

- Fix bug in join config parameter that rendered it unusable from config file.

## 0.5.1 (2015-11-06)

- Deb package
- Upgraded libkv to latest
- New config options (log level, web UI path)

## 0.5.0 (2015-09-27)

- Email and Webhook configurable notifications for job executions.
- Ability to encrypt serf network traffic between nodes.
- Pretty formating API responses
- UI now shows the execution status with color coding and partial execution.
- More API stability and predictability
- Provided API JSON schema, generated API docs based in the schema
- Tested on Travis
- Using Libkv allows to use different storage backends (etcd, consul, zookeeper)
- Add v1 versioning to the API routes

## 0.0.4 (2015-08-23)

- Compiled with Go 1.5
- Includes cluster nodes view in the UI

## 0.0.3 (2015-08-20)

- Initial release


dkron/config.go
package dkron

import (
	"encoding/base64"
	"flag"
	"fmt"
	"net"
	"os"
	"time"

	"github.com/Sirupsen/logrus"
	"github.com/spf13/viper"
)

type Config struct {
	NodeName              string
	BindAddr              string
	HTTPAddr              string
	Discover              string
	Backend               string
	BackendMachines       []string
	Profile               string
	Interface             string
	AdvertiseAddr         string
	Tags                  map[string]string
	SnapshotPath          string
	ReconnectInterval     time.Duration
	ReconnectTimeout      time.Duration
	TombstoneTimeout      time.Duration
	DisableNameResolution bool
	KeyringFile           string
	RejoinAfterLeave      bool
	Server                bool
	EncryptKey            string
	StartJoin             AppendSliceValue
	Keyspace              string
	UIDir                 string
	RPCPort               int
	AdvertiseRPCPort      int

	MailHost     string
	MailPort     uint16
	MailUsername string
	MailPassword string
	MailFrom     string
	MailPayload  string

	WebhookURL     string
	WebhookPayload string
	WebhookHeaders []string

	// DogStatsdAddr is the address of a dogstatsd instance. If provided,
	// metrics will be sent to that instance
	DogStatsdAddr string
	// DogStatsdTags are the global tags that should be sent with each packet to dogstatsd
	// It is a list of strings, where each string looks like "my_tag_name:my_tag_value"
	DogStatsdTags []string
	StatsdAddr    string
}

// This is the default port that we use for Serf communication
const DefaultBindPort int = 8946

func init() {
	viper.SetConfigName("dkron")        // name of config file (without extension)
	viper.AddConfigPath("/etc/dkron")   // call multiple times to add many search paths
	viper.AddConfigPath("$HOME/.dkron") // call multiple times to add many search paths
	viper.AddConfigPath("./config")     // call multiple times to add many search paths
	viper.SetEnvPrefix("dkron")         // will be uppercased automatically
	viper.AutomaticEnv()
}

// readConfig is responsible for setup of our configuration using
// the command line and any file configs
func NewConfig(args []string, agent *AgentCommand) *Config {
	hostname, err := os.Hostname()
	if err != nil {
		log.Panic(err)
	}

	cmdFlags := flag.NewFlagSet("server", flag.ContinueOnError)
	cmdFlags.Usage = func() { agent.Ui.Output(agent.Help()) }
	cmdFlags.String("node", hostname, "node name")
	viper.SetDefault("node_name", cmdFlags.Lookup("node").Value)
	cmdFlags.String("bind", fmt.Sprintf("0.0.0.0:%d", DefaultBindPort), "address to bind listeners to")
	viper.SetDefault("bind_addr", cmdFlags.Lookup("bind").Value)
	cmdFlags.String("advertise", "", "address to advertise to other nodes")
	viper.SetDefault("advertise_addr", cmdFlags.Lookup("advertise").Value)
	cmdFlags.String("http-addr", ":8080", "HTTP address")
	viper.SetDefault("http_addr", cmdFlags.Lookup("http-addr").Value)
	cmdFlags.String("discover", "dkron", "mDNS discovery name")
	viper.SetDefault("discover", cmdFlags.Lookup("discover").Value)
	cmdFlags.String("backend", "etcd", "store backend")
	viper.SetDefault("backend", cmdFlags.Lookup("backend").Value)
	cmdFlags.String("backend-machine", "127.0.0.1:2379", "store backend machines addresses")
	viper.SetDefault("backend_machine", cmdFlags.Lookup("backend-machine").Value)
	cmdFlags.String("profile", "lan", "timing profile to use (lan, wan, local)")
	viper.SetDefault("profile", cmdFlags.Lookup("profile").Value)
	viper.SetDefault("server", cmdFlags.Bool("server", false, "start dkron server"))
	var startJoin []string
	cmdFlags.Var((*AppendSliceValue)(&startJoin), "join", "address of agent to join on startup")
	var tag []string
	cmdFlags.Var((*AppendSliceValue)(&tag), "tag", "tag pair, specified as key=value")
	cmdFlags.String("keyspace", "dkron", "key namespace to use")
	viper.SetDefault("keyspace", cmdFlags.Lookup("keyspace").Value)
	cmdFlags.String("encrypt", "", "encryption key")
	viper.SetDefault("encrypt", cmdFlags.Lookup("encrypt").Value)

	cmdFlags.String("log-level", "info", "Log level (debug, info, warn, error, fatal, panic), defaults to info")
	viper.SetDefault("log_level", cmdFlags.Lookup("log-level").Value)

	viper.SetDefault("rpc_port", cmdFlags.Int("rpc-port", 6868, "RPC port"))
	viper.SetDefault("advertise_rpc_port", cmdFlags.Int("advertise-rpc-port", 0, "advertise RPC port"))

	// Notifications
	cmdFlags.String("mail-host", "", "notification mail server host")
	viper.SetDefault("mail_host", cmdFlags.Lookup("mail-host").Value)
	cmdFlags.String("mail-port", "", "port to use for the mail server")
	viper.SetDefault("mail_port", cmdFlags.Lookup("mail-port").Value)
	cmdFlags.String("mail-username", "", "username for the mail server")
	viper.SetDefault("mail_username", cmdFlags.Lookup("mail-username").Value)
	cmdFlags.String("mail-password", "", "password of the mail server")
	viper.SetDefault("mail_password", cmdFlags.Lookup("mail-password").Value)
	cmdFlags.String("mail-from", "", "notification emails from address")
	viper.SetDefault("mail_from", cmdFlags.Lookup("mail-from").Value)
	cmdFlags.String("mail-payload", "", "notification mail payload")
	viper.SetDefault("mail_payload", cmdFlags.Lookup("mail-payload").Value)

	cmdFlags.String("webhook-url", "", "notification webhook url")
	viper.SetDefault("webhook_url", cmdFlags.Lookup("webhook-url").Value)
	cmdFlags.String("webhook-payload", "", "notification webhook payload")
	viper.SetDefault("webhook_payload", cmdFlags.Lookup("webhook-payload").Value)
	webhookHeaders := &AppendSliceValue{}
	cmdFlags.Var(webhookHeaders, "webhook-header", "notification webhook additional header")

	cmdFlags.String("dog-statsd-addr", "", "DataDog Agent address")
	viper.SetDefault("dog_statsd_addr", cmdFlags.Lookup("dog-statsd-addr").Value)
	var dogStatsdTags []string
	cmdFlags.Var((*AppendSliceValue)(&dogStatsdTags), "dog-statsd-tags", "Datadog tags, specified as key:value")
	cmdFlags.String("statsd-addr", "", "Statsd Address")
	viper.SetDefault("statsd_addr", cmdFlags.Lookup("statsd-addr").Value)

	if err := cmdFlags.Parse(args); err != nil {
		log.Fatal(err)
	}

	// Set array params defaults
	ut, err := UnmarshalTags(tag)
	if err != nil {
		log.Fatal(err)
	}
	viper.SetDefault("tags", ut)
	viper.SetDefault("join", startJoin)
	viper.SetDefault("webhook_headers", webhookHeaders)
	viper.SetDefault("dog_statsd_tags", dogStatsdTags)

	return ReadConfig(agent)
}

func ReadConfig(agent *AgentCommand) *Config {
	err := viper.ReadInConfig() // Find and read the config file
	if err != nil {             // Handle errors reading the config file
		logrus.WithError(err).Info("No valid config found: Applying default values.")
	}

	tags := viper.GetStringMapString("tags")
	server := viper.GetBool("server")
	nodeName := viper.GetString("node_name")

	if server {
		tags["dkron_server"] = "true"
	}
	tags["dkron_version"] = agent.Version

	InitLogger(viper.GetString("log_level"), nodeName)

	return &Config{
		NodeName:         nodeName,
		BindAddr:         viper.GetString("bind_addr"),
		AdvertiseAddr:    viper.GetString("advertise_addr"),
		HTTPAddr:         viper.GetString("http_addr"),
		Discover:         viper.GetString("discover"),
		Backend:          viper.GetString("backend"),
		BackendMachines:  viper.GetStringSlice("backend_machine"),
		Server:           server,
		Profile:          viper.GetString("profile"),
		StartJoin:        viper.GetStringSlice("join"),
		Tags:             tags,
		Keyspace:         viper.GetString("keyspace"),
		EncryptKey:       viper.GetString("encrypt"),
		UIDir:            viper.GetString("ui_dir"),
		RPCPort:          viper.GetInt("rpc_port"),
		AdvertiseRPCPort: viper.GetInt("advertise_rpc_port"),

		MailHost:     viper.GetString("mail_host"),
		MailPort:     uint16(viper.GetInt("mail_port")),
		MailUsername: viper.GetString("mail_username"),
		MailPassword: viper.GetString("mail_password"),
		MailFrom:     viper.GetString("mail_from"),
		MailPayload:  viper.GetString("mail_payload"),

		WebhookURL:     viper.GetString("webhook_url"),
		WebhookPayload: viper.GetString("webhook_payload"),
		WebhookHeaders: viper.GetStringSlice("webhook_headers"),

		DogStatsdAddr: viper.GetString("dog_statsd_addr"),
		DogStatsdTags: viper.GetStringSlice("dog_statsd_tags"),
		StatsdAddr:    viper.GetString("statsd_addr"),
	}
}

// AddrParts returns the parts of the BindAddr that should be
// used to configure Serf.
func (c *Config) AddrParts(address string) (string, int, error) {
	checkAddr := address

START:
	_, _, err := net.SplitHostPort(checkAddr)
	if ae, ok := err.(*net.AddrError); ok && ae.Err == "missing port in address" {
		checkAddr = fmt.Sprintf("%s:%d", checkAddr, DefaultBindPort)
		goto START
	}
	if err != nil {
		return "", 0, err
	}

	// Get the address
	addr, err := net.ResolveTCPAddr("tcp", checkAddr)
	if err != nil {
		return "", 0, err
	}

	return addr.IP.String(), addr.Port, nil
}

// Networkinterface is used to get the associated network
// interface from the configured value
func (c *Config) NetworkInterface() (*net.Interface, error) {
	if c.Interface == "" {
		return nil, nil
	}
	return net.InterfaceByName(c.Interface)
}

// EncryptBytes returns the encryption key configured.
func (c *Config) EncryptBytes() ([]byte, error) {
	return base64.StdEncoding.DecodeString(c.EncryptKey)
}
